CONTRIBUTING.md

```diff
@@ -0,0 +1,76 @@
+this operator.
+
+- Generally, before developing enhancements to this charm, you should consider [opening an issue
+  ](https://github.com/canonical/training-operator/issues) explaining your use case.
+- If you would like to chat with us about your use-cases or proposed implementation, you can reach
+  us at [Canonical Mattermost public channel](https://chat.charmhub.io/charmhub/channels/charm-dev)
+  or [Discourse](https://discourse.charmhub.io/).
+- Familiarising yourself with the [Charmed Operator Framework](https://juju.is/docs/sdk) library
+  will help you a lot when working on new features or bug fixes.
+- All enhancements require review before being merged. Code review typically examines
+  - code quality
+  - test coverage
+  - user experience for Juju administrators of this charm.
+- Please help us out in ensuring easy to review branches by rebasing your pull request branch onto
+  the `main` branch. This also avoids merge commits and creates a linear Git commit history.
+
+## Developing
+
+You can use the environments created by `tox` for development:
+
+```shell
+tox --notest -e unit
+source .tox/unit/bin/activate
+```
+
+### Testing
+
+```shell
+tox -e lint          # code style
+tox -e unit          # unit tests
+tox -e integration   # integration tests
+tox                  # runs 'lint' and 'unit' environments
+```
+
+## Build charm
+
+Build the charm in this git repository using:
+
+```shell
+charmcraft pack
+```
+
+### Deploy
+
+```bash
+# Create a model
+juju add-model dev
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm
+juju deploy ./training-operator_ubuntu-20.04-amd64.charm \
+    --resource training-operator-image=kubeflow/training-operator
+```
+
+## Charmed Training Operator source and resources upgrade
+
+To upgrade the source and resources Charmed Training Operator, you must:
+
+1) Bump the `training-operator-image` in `metadata.yaml`
+2) Update all manifests in `src/` to match those in the version you want to upgrade to
+3) Ensure integration and unit tests are passing; fix/adapt them otherwise
+
+The following table collects a series of links and information that can help in the upgrade process.
+ 
+| Image repository 	| Image version 	| Manifest location 	| Next version 	|
+|---	|---	|---	|---	|
+| [kubeflow/training-operator](https://hub.docker.com/r/kubeflow/training-operator/) 	| `v1-<commit sha>`<br>The commit sha comes from the HEAD of the release tag 	| [kubeflow/training-operator/manifests](https://github.com/kubeflow/training-operator/tree/v1.5.0-rc.0/manifests) 	| Check for latest version released or pre-released in <br>the repo [releases](https://github.com/kubeflow/training-operator/releases), it should match the release plan usually stated in issues like <br>[Training Operator WG and Kubeflow 1.6](https://github.com/kubeflow/manifests/issues/2196) release and [KF Release 1.6 Tracking](https://github.com/kubeflow/manifests/issues/2194) 	|
```

Comment:
re conversation about `kubeflow/manifests` vs workload repo:
> My reasoning is: kubeflow/manifests is almost always behind the workload repo, they usually tag releases (or release candidates) about two weeks after the workload release (or pre-release).  ....

Yeah I see what you mean.  I've been thinking of this as "our KF 1.x release should be in line with KF's 1.x release", and then to me `kubeflow/manifests` makes the most sense as the source of truth.  At minimum I agree sometimes we might want to deviate from that, but I'm not sure we want to deviate by default.  I worry for applications that release quickly.  There have been times in the past where some components (I think it was kfp?) did not put their newest release into the current Kubeflow (can't remember why.  might have just been a timing thing?), so I worry that if our contributing nudges people to the component's repo first it'll make an accident easier.  Configuration could also be different between the different repos, although I don't know if that actually happens much

What do think about having links here to both:
* `kubeflow/manifests/whateverDirThisAppIsIn` (direct to the dir so people don't have to search that big dir tree)
* `thisrepo/manifests`

And instructions that make it clear that we typically sync our releases with those in kubeflow/manifests, but sometimes the repo/manifests are also helpful?

Reply:
> I worry for applications that release quickly.

What worries you the most about this? They always tag their releases (and RCs).

> There have been times in the past where some components (I think it was kfp?) did not put their newest release into the current Kubeflow (can't remember why. might have just been a timing thing?), so I worry that if our contributing nudges people to the component's repo first it'll make an accident easier. 

What do you mean by "put their newest release into the current Kubeflow"?

> And instructions that make it clear that we typically sync our releases with those in kubeflow/manifests, but sometimes the repo/manifests are also helpful?

This works for me.

Reply:
Agreed that this is a charm not the bundle repo, but in practice right now how `kubeflow/manifests` applies the application is more important to us than how `repo/manifests` does.  Although they're similar, they're often not the same.  @DomFleischmann are you against having both?  Knowing where to find this charm in `kubeflow/manifests` is pretty valuable as it is a pretty complicated dir tree

Reply:
#### Re worries about applications with quick releases (@DnPlas): 

yes, the images are tagged, but if we're looking at `repo/manifests` we have no context of which tag we should look for.  Generally, we're trying to sync with what is being released in `kubeflow/manifests @ v1.X`, which is information only available in `kubeflow/manifests`

#### Re "put their newest release into the current Kubeflow": 

By that I mean a repo may have released X.Z as their newest release, but Kubeflow is still using X.Y.  Looking at `repo/manifests` doesn't tell us that, it just points us toward how to configure X.Z.

Reply:
I like the idea of putting both `kubeflow/manifests` and `repo/manifests`, let's do that and resolve this conversation.